README.md
# i_mth

This is a crate designed for use in easy to complex statics and dynamics problems.

The crate is made as lean as possible and method names are directly correlated to what they
actually do.

All types currently in the crate and features that will be implemented later will all use the
`f64` data type because well I am a freak of nature and I have accuracy issues.

## Usage

Add this to `Cargo.toml`

```toml
[dependencies]
i_mth = "0.1.2"
```

## Current Crate available features

1. Support for both 2D and 3D vectors.
1. Common Constants.
1. A utils module for calculating escape velocity and acceleration of gravity of other celestial bodies. (other functions will be added in the future).

## Currently in the workings

1. Support For matrices, operations on matrices and transformations
1. Numerical Calculus

### Examples

1. Creating 2D and 3D vectors with i_mth

```rust
    use i_mth::vector3d::Vector3D;

    fn main() {
        // there are several available methods to create vectors
        let v3d_0 = Vector3D::new(1.0, 1.0, 1.0);
        let v3d_1 = Vector3D::set(1.0);

        assert_eq!(v3d_0, v3d_1);
    }
```

2. Moment of a force About a **O**: The moment of a force **F** about  **O** can be defined as the vector product (cross product) of **r** and **F**.
Where **r** is the position vector between the point of application of the force to the fixed reference point **O**.
ie **Moment** = **r x F**

```rust
    use i_mth::vector3d::Vector3D;

    fn main() {
        let f = Vector3D::new(400.0, 693.0, 0.0);
        let r = Vector3D::new(-0.2, 0.16, 0.0);

        let moment = r.cross(f);
        assert_eq!(-202.6, moment.z);
    }
```

3. Using the utils to calculate [escape velocity](https://en.wikipedia.org/wiki/Escape_velocity#:~:text=More%20generally%2C%20escape%20velocity%20is,orbit%20(of%20any%20radius).): Escape velocity or escape speed is the minimum speed needed for a free, non-propelled object to escape from the gravitational influence of a primary body, thus reaching an infinite distance from it.

```rust
    use i_mth::utils::calc_escape_velocity;

    fn main() {
        let mass_of_moon = 7.342e22;
        let radius_of_moon = 1737.4e3;

        let escape_vel_of_moon = calc_escape_velocity(mass_of_moon, radius_of_moon);

        println!("{} km/s", escape_vel_of_moon / 1000.0);
    }
```

### Supports fmt

```rust
    use i_mth::vector3d::Vector3D;

    fn main() {
        let unit_i = Vector3D::i();
        let unit_j = Vector3D::j();

        let unit_k = unit_i.cross(unit_j);

        println!("{}", unit_k);
    }
```


CHANGELOG.md
# Changelog

## 0.1.3 -> Currently in the working

- Support For matrices, operations on matrices, rotations of vectors and vector transformations
- numerical calculus

## 0.1.2

- Fixed the error in the version number in README.md
- Fixed some documentations errors in constants.rs

## 0.1.1

- fixed path errors

## 0.1.0

- Added constants for common mathematical and physical constants.
- Fixed some non-uniform syntax between Self and Vector3D
- Added utils.rs
- Finally learned about semantic versioning
- Using semantic versioning

## 0.0.2

- fixed a silly doc-test error

## 0.0.1

- Created the basic layout of the vector module

## 0.0.0

- Initial Project Structure with more or less no files and code.


src/constants.rs
//! # CONSTANTS
//!
//! Stores common mathematical and physical constants as well as
//!
//! Again each constant is stored as an f64 type and uncertainty of these
//! constants are specified in the links.
//!
//! [Sources for mathematical constants](https://en.wikipedia.org/wiki/List_of_mathematical_constants)
//! [Sources for physical constants](https://en.wikipedia.org/wiki/List_of_physical_constants)
//! [More Sources](https://cosmologist.info/teaching/Cosmology/Physical_constants.pdf)


/// **Acceleration due to gravity**\
/// unit: ms<sup>-2</sup>
/// value: -9.806_65
pub const EARTH_GRAVITY: f64 = -9.806_65;

/// **Mass of planet Earth**
/// unit: kg
/// value: 5.972_168e24
/// [More Info](https://en.wikipedia.org/wiki/Earth)
pub const EARTH_MASS: f64 = 5.972_168e24;

/// **Radius of planet Earth**
/// unit: m
/// value: 6_371e3
/// [More Info](https://en.wikipedia.org/wiki/Earth)
pub const EARTH_RADIUS: f64 = 6_371e3;

/// **Newtonian gravitational constant**\
/// unit: m<sup>3</sup>kg<sup>-1</sup>s<sup>-2<sup>\
/// value: 6.674_30e-11
/// [More Info](https://en.wikipedia.org/wiki/Gravitational_constant)
pub const G: f64 = 6.674_30e-11;

/// **The constant PI**\
/// unit: dimensionless\
/// value: 3.14159_26535_89793
/// [More Info](https://en.wikipedia.org/wiki/Pi)
pub const PI: f64 = 3.14159_26535_89793;

/// **The constant Tau**
/// value: 6.28318_53071_79586
/// [More Info](https://tauday.com/tau-manifesto)
pub const TAU: f64 = 6.28318_53071_79586;

/// **The Speed of light in a vacuum**\
/// unit: ms<sup>-1</sup>\
/// value: 299_792_458
/// [More Info](https://en.wikipedia.org/wiki/Speed_of_light)
pub const C: usize = 299_792_458;

/// **Euler's Number**\
/// value: 2.718_281_828_459_045
/// [More Info](https://en.wikipedia.org/wiki/E_(mathematical_constant))
pub const E: f64 = 2.718_281_828_459_045;

/// **Vacuum Permeability**\
/// unit: m.kg.s<sup>-2</sup>A<sup>-2</sup>\
/// value: 1.256_637_06e-6
/// [More Info](https://en.wikipedia.org/wiki/Vacuum_permeability)
pub const VACUUM_PERMEABILITY: f64 = 1.256_637_06e-6;

/// **Vacuum Permittivity**\
/// unit: Fm<sup>-1</sup>\
/// value: 8.854_187_812_8e-12
/// [More Info](https://en.wikipedia.org/wiki/Vacuum_permittivity)
pub const VACUUM_PERMITTIVITY: f64 = 8.854_187_812_8e-12;

src/vectors/vector2d.rs
use std::fmt;
use std::ops::*;
use crate::vector3d::Vector3D;

/// Represents a mathematical vector in 2 dimensional space.
#[derive(Debug, PartialEq, Clone, Copy, Default)]
#[repr(C)]
pub struct Vector2D {
    pub x: f64,
    pub y: f64,
}

impl Vector2D {
    /// Returns a new vector with the components passed.\
    /// x is in the direction of the i-unit vector.\
    /// y is in the direction of the j-unit vector.
    ///
    /// # Example
    ///
    /// ```rust
    /// use i_mth::vector2d::Vector2D;
    ///
    /// let vec2d = Vector2D::new(1.0, 1.0);
    ///
    /// assert_eq!(1.0, vec2d.x);
    /// assert_eq!(1.0, vec2d.y);
    /// ```
    #[inline]
    pub fn new(x: f64, y: f64) -> Self {
        Self { x, y }
    }

    /// Creates a new vector with the x and y components set to the value passed.
    ///
    /// # Example
    /// ```rust
    /// use i_mth::vector2d::Vector2D;
    ///
    /// let vec2d = Vector2D::set(1.0);
    ///
    /// assert_eq!(1.0, vec2d.x);
    /// assert_eq!(1.0, vec2d.y);
    /// ```
    #[inline]
    pub fn set(value: f64) -> Self {
        Self { x: value, y: value }
    }
    /// Return the unit vector i == (i + 0j)
    #[inline]
    pub fn i() -> Self {
        Self { x: 1.0, y: 0.0 }
    }

    /// Returns the unit vector j == (0i + j)
    #[inline]
    pub fn j() -> Self {
        Self { x: 0.0, y: 1.0 }
    }

    /// Returns a vector pointing to the origin of the coordinate system
    /// (0i + 0j)
    #[inline]
    pub fn origin() -> Self {
        Self { x: 0.0, y: 0.0 }
    }

    /// Returns a vector with the selected component set to the passed value,
    /// while other components gets set to zero.
    /// If an invalid component label like "a" is selected, None is returned.
    /// 
    /// Valid component labels are i, j, or x, y
    /// 
    /// # Example
    /// ```rust
    /// use i_mth::vector2d::Vector2D;
    /// 
    /// // "j" can be used instead of y
    /// let acc_due_to_gravity = Vector2D::select("y", -9.81);
    /// assert_eq!(-9.81, acc_due_to_gravity.unwrap().y);
    #[inline]
    pub fn select(comp: &str, value: f64) -> Option<Vector2D> {
        match comp {
            "i" | "x" => Some(Vector2D { x: value, y: 0.0 }),
            "j" | "y" => Some(Vector2D { x: 0.0, y: value }),
            _ => None,
        }
    }

    /// Returns the dot product of this vector and the passed vector
    #[inline]
    pub fn dot(&self, other: Vector2D) -> f64 {
        (self.x * other.x) + (self.y * other.y)
    }

    /// Multiples the x, y, and z components of this vector by the x, y, z components
    /// of the passed vector.
    #[inline]
    pub fn scale_comps_by_comps(&self, other: Vector2D) -> Self {
        Self {
            x: self.x * other.x,
            y: self.y * other.y,
        }
    }

    /// Returns a vector with the absolute values of this vectors components
    #[inline]
    pub fn abs(&self) -> Self {
        Self {
            x: self.x.abs(),
            y: self.y.abs(),
        }
    }

    /// Returns a vector with this vector's components scaled by the passed value
    #[inline]
    pub fn scale(&self, value: f64) -> Self {
        Self {
            x: self.x * value,
            y: self.y * value,
        }
    }

    /// Converts this vector to its unit form if arithmetically possible else
    /// Returns this vector unchanged if it fails. This operation can fail
    /// if you have a zero vector
    #[inline]
    pub fn to_unit(&mut self) {
        let mag = self.magnitude();
        if mag > 0.0 {
            let inv_mag = 1.0 / mag;
            self.x *= inv_mag;
            self.y *= inv_mag;
        }
    }

    /// Returns the normalized(unit) version of this vector if arithmetically possible
    /// else it returns None. This operation can fail if you have a zero vector.
    #[inline]
    pub fn normalized(&self) -> Option<Self> {
        let mag = self.magnitude();
        if mag > 0.0 {
            let inv_mag = 1.0 / mag;
            return Some(Vector2D {
                x: self.x * inv_mag,
                y: self.y * inv_mag,
            });
        }
        None
    }

    /// Scales the passed vector by the passed value and performs vector
    /// addition on this vector and the other vector.
    #[inline]
    pub fn add_scaled(self, other: Vector2D, value: f64) -> Self {
        self + other.scale(value)
    }

    /// Scales this vector by the passed value and performs vector
    /// addition on this vector and the passed vector.
    #[inline]
    pub fn scale_add(&self, value: f64, other: Vector2D) -> Self {
        self.scale(value) + other
    }

    /// Returns true if this vector is equal to the passed vector.
    #[inline]
    pub fn is_equal_to(&self, other: Vector2D) -> bool {
        (self.x == other.x) && (self.y == other.y)
    }

    /// Returns true if this vector has a greater magnitude(length) than the passed vector.
    #[inline]
    pub fn is_greater_than(&self, other: Vector2D) -> bool {
        self.squared_magnitude() > other.squared_magnitude()
    }

    /// If the components are matched x to x, y to y, and z to z, the method returns true only
    /// if every component of this vector is greater than every component of the passed vector.
    #[inline]
    pub fn comp_wise_gt(&self, other: Vector2D) -> bool {
        (self.x > other.x) && (self.y > other.y)
    }

    /// Converts this vector from cartesian to cylindrical components
    #[inline]
    pub fn as_cylindrical(&self) -> Self {
        Self {
            x: ((self.x * self.x) + (self.y * self.y)).sqrt(),
            y: (self.y / self.x).atan(),
        }
    }

    /// Converts this vector to a 3 Dimensional one by the addition of the
    /// z component passed
    pub fn to_3d(&self, z: f64) -> Vector3D {
        Vector3D {
            x: self.x,
            y: self.y,
            z,
        }
    }
    /// Returns the squared magnitude of this vector.
    pub fn squared_magnitude(&self) -> f64 {
        (self.x * self.x) + (self.y * self.y)
    }
    /// Returns the magnitude of this vector.
    #[inline]
    pub fn magnitude(&self) -> f64 {
        self.squared_magnitude().sqrt()
    }
}

impl Add for Vector2D {
    type Output = Self;
    #[inline]
    fn add(self, rhs: Self) -> Self {
        Self {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
        }
    }
}

impl AddAssign for Vector2D {
    #[inline]
    fn add_assign(&mut self, rhs: Self) {
        self.x += rhs.x;
        self.y += rhs.y;
    }
}

impl Sub for Vector2D {
    type Output = Self;
    #[inline]
    fn sub(self, rhs: Self) -> Self {
        Self {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
        }
    }
}

impl SubAssign for Vector2D {
    #[inline]
    fn sub_assign(&mut self, rhs: Self) {
        self.x -= rhs.x;
        self.y -= rhs.y;
    }
}

impl Mul for Vector2D {
    type Output = Self;
    #[inline]
    fn mul(self, rhs: Self) -> Self {
        Self {
            x: self.x * rhs.x,
            y: self.y * rhs.y,
        }
    }
}

impl MulAssign for Vector2D {
    #[inline]
    fn mul_assign(&mut self, rhs: Self) {
        self.x *= rhs.x;
        self.y *= rhs.y;
    }
}

impl Div for Vector2D {
    type Output = Self;
    #[inline]
    fn div(self, rhs: Self) -> Self {
        Self {
            x: self.x / rhs.x,
            y: self.y / rhs.y,
        }
    }
}

impl DivAssign for Vector2D {
    #[inline]
    fn div_assign(&mut self, rhs: Self) {
        self.x /= rhs.x;
        self.y /= rhs.y;
    }
}

impl fmt::Display for Vector2D {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}i + {}j", self.x, self.y)
    }
}

impl fmt::Binary for Vector2D {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let magnitude = self.magnitude();
        let decimals = f.precision().unwrap_or(4);
        let string = format!("{magnitude:.decimals$}");
        f.pad_integral(true, "", &string)
    }
}

src/vectors/vector3d.rs
use std::fmt;
use std::ops::*;
use crate::vector2d::Vector2D;

/// Represents a mathematical vector in 3 Dimensional space.
#[derive(Debug, PartialEq, Clone, Copy, Default)]
#[repr(C)]
pub struct Vector3D {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

impl Vector3D {

    /// Returns a new vector with the passed components.\
    /// x is in the direction of the i-unit vector.\
    /// y is in the direction of the j-unit vector.\
    /// z is in the direction of the k-unit vector.
    /// 
    /// # Example
    /// 
    /// ```rust
    /// use i_mth::vector3d::Vector3D;
    /// 
    /// let vec3d = Vector3D::new(1.0, 1.0, 1.0);
    /// 
    /// assert_eq!(1.0, vec3d.x);
    /// assert_eq!(1.0, vec3d.y);
    /// assert_eq!(1.0, vec3d.z);
    /// ```
    #[inline]
    pub fn new(x: f64, y: f64, z: f64) -> Self {
        Self { x, y, z }
    }

    /// Creates a new vector with the x, y, and z components set to the value passed.
    /// 
    /// # Example
    /// ```rust
    /// use i_mth::vector3d::Vector3D;
    /// 
    /// let vec3d = Vector3D::set(1.0);
    /// 
    /// assert_eq!(1.0, vec3d.x);
    /// assert_eq!(1.0, vec3d.y);
    /// assert_eq!(1.0, vec3d.z);
    /// ```
    #[inline]
    pub fn set(value: f64) -> Self {
        Self {
            x: value,
            y: value,
            z: value,
        }
    }
    /// Return the unit vector i == (i + 0j + 0k)
    #[inline]
    pub fn i() -> Self {
        Self { x: 1.0, y: 0.0, z: 0.0 }
    }

    #[inline]
    /// Returns the unit vector j == (0i + j + 0k)
    pub fn j() -> Self {
        Self { x: 0.0, y: 1.0, z: 0.0 }
    }

    /// Returns the unit vector k == (0i + 0j + k)
    #[inline]
    pub fn k() -> Self {
        Self { x: 0.0, y: 0.0, z: 1.0 }
    }

    /// Returns a vector pointing to the origin of the coordinate system
    /// (0i + 0j + 0k)
    #[inline]
    pub fn origin() -> Self {
        Self { x: 0.0, y: 0.0, z: 0.0 }
    }
    
    /// Returns a vector with the selected component set to the passed value,
    /// while other components gets set to zero.
    /// If an invalid component label like "a" is selected, None is returned.
    /// 
    /// Valid component labels are i, j, k or x, y, z
    /// 
    /// # Example
    /// ```rust
    /// use i_mth::vector3d::Vector3D;
    /// 
    /// // "j" can be used instead of y
    /// let acc_due_to_gravity = Vector3D::select("y", -9.81);
    /// assert_eq!(-9.81, acc_due_to_gravity.unwrap().y);
    #[inline]
    pub fn select(comp: &str, value: f64) -> Option<Vector3D> {
        match comp {
            "i" | "x" => Some(Vector3D{x: value, y: 0.0, z: 0.0}),
            "j" | "y" => Some(Vector3D{x: 0.0, y: value, z: 0.0}),
            "k" | "z" => Some(Vector3D{x: 0.0, y: 0.0, z: value}),
                _     => None,
        }
    }
    /// Returns the dot product of this vector and the passed vector
    #[inline]
    pub fn dot(&self, other: Vector3D) -> f64 {
        (self.x * other.x) + (self.y * other.y) + (self.z * other.z) 
    }

    /// Returns the cross product of this vector and the passed vector
    #[inline]
    pub fn cross(&self, other: Vector3D) -> Self {
        Self {
            x: ((self.y * other.z) + (-self.z * other.y)),
            y: ((self.z * other.x) + (-self.x * other.z)),
            z: ((self.x * other.y) + (-self.y * other.x)),
        }
    }

    /// Multiples the x, y, and z components of this vector by the x, y, z components
    /// of the passed vector.
    #[inline]
    pub fn scale_comps_by_comps(&self, other: Vector3D) -> Self {
        Self {
            x: self.x * other.x,
            y: self.y * other.y,
            z: self.z * other.z,
        }
    }

    /// Returns the triple scalar product of this vector and the passed vectors.
    /// ### NOTE: the arrangement is -> this_vector . (oth_1 x oth_2)
    #[inline]
    pub fn triple_scalar_prod(&self, oth_1: Vector3D, oth_2: Vector3D) -> f64 {
        self.x * ((oth_1.y * oth_2.z) - (oth_1.z * oth_2.y)) +
        self.y * ((oth_1.z * oth_2.x) - (oth_1.x * oth_2.z)) + 
        self.z * ((oth_1.x * oth_2.y) - (oth_1.y * oth_2.x))
    }

    /// Returns the triple vector product of this vector and the passed vectors.
    /// ### NOTE: the arrangement is -> this_vector x (oth_1 x oth_2)
    #[inline]
    pub fn triple_vector_prod(&self, oth_1: Vector3D, oth_2: Vector3D) -> Self {
        let temp = oth_1.cross(oth_2);
        self.cross(temp)
    }

    /// Returns the squared magnitude of this vector.
    #[inline]
    pub fn squared_magnitude(&self) -> f64 {
        (self.x * self.x) + (self.y * self.y) + (self.z * self.z)
    }

    /// Returns the magnitude of this vector.
    #[inline]
    pub fn magnitude(&self) -> f64 {
        self.squared_magnitude().sqrt()
    }

    /// Returns a vector with the absolute values of this vectors components
    #[inline]
    pub fn abs(&self) -> Self {
        Self {
            x: self.x.abs(),
            y: self.y.abs(),
            z: self.z.abs(),
        }
    }

    /// Returns a vector with this vector's components scaled by the passed value
    #[inline]
    pub fn scale(&self, value: f64) -> Self {
        Self {
            x: self.x * value,
            y: self.y * value,
            z: self.z * value,
        }
    }

    /// Converts this vector to its unit form if arithmetically possible else
    /// Returns this vector unchanged if it fails. This operation can fail
    /// if you have a zero vector
    #[inline]
    pub fn to_unit(&mut self) {
        let mag = self.magnitude();
        if mag > 0.0 {
            let inv_mag = 1.0 / mag;
            self.x *= inv_mag;
            self.y *= inv_mag;
            self.z *= inv_mag;
        }
    }

    /// Returns the normalized(unit) version of this vector if arithmetically possible
    /// else it returns None. This operation can fail if you have a zero vector.
    #[inline]
    pub fn normalized(&self) -> Option<Self> {
        let mag = self.magnitude();
        if mag > 0.0 {
            let inv_mag = 1.0 / mag;
            return Some(Vector3D {
                x: self.x * inv_mag,
                y: self.y * inv_mag,
                z: self.z * inv_mag,
            })
        }
        None
    }

    /// Scales the passed vector by the passed value and performs vector
    /// addition on this vector and the other vector.
    #[inline]
    pub fn add_scaled(self, other: Vector3D, value: f64) -> Self {
        self + other.scale(value)
    }

    /// Scales this vector by the passed value and performs vector
    /// addition on this vector and the passed vector.
    #[inline]
    pub fn scale_add(&self, value: f64, other: Vector3D) -> Self {
        self.scale(value) + other
    }
    
    /// Returns a 2 Dimensional vector by truncating the z component of this vector.
    #[inline]
    pub fn to_2d(&self) -> Vector2D {
        Vector2D {
            x: self.x,
            y: self.y,
        }
    }

    /// Returns true if this vector is equal to the passed vector.
    #[inline]
    pub fn is_equal_to(&self, other: Vector3D) -> bool {
        (self.x == other.x) && (self.y == other.y) && (self.z == other.z)
    }

    /// Returns true if this vector has a greater magnitude(length) than the passed vector. 
    #[inline]
    pub fn is_greater_than(&self, other: Vector3D) -> bool {
        self.squared_magnitude() > other.squared_magnitude()
    }

    /// If the components are matched x to x, y to y, and z to z, the method returns true only
    /// if every component of this vector is greater than every component of the passed vector.
    #[inline]
    pub fn comp_wise_gt(&self, other: Vector3D) -> bool {
        (self.x > other.x) && (self.y > other.y) && (self.z > other.z)
    }

    /// Converts this vector from cartesian to cylindrical components
    #[inline]
    pub fn as_cylindrical(&mut self) {
        self.x = ((self.x * self.x) + (self.y * self.y)).sqrt();
        self.y = (self.y/ self.x).atan();

    }

    /// Converts this vector from cartesian to spherical components
    #[inline]
    pub fn as_spherical(&mut self) {
        self.x = self.magnitude();
        self.y = (self.z / self.magnitude()).acos();
        self.z = (self.y / self.x).atan();

    }
}


impl Add for Vector3D {
    type Output = Self;
    #[inline]
    fn add(self, rhs: Self) -> Self {
        Self {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
            z: self.z + rhs.z,
        }
    }

}

impl AddAssign for Vector3D {
    #[inline]
    fn add_assign(&mut self, rhs: Self) {
        self.x += rhs.x;
        self.y += rhs.y;
        self.z += rhs.z;
    }
}

impl Sub for Vector3D {
    type Output = Self;
    #[inline]
    fn sub(self, rhs: Self) -> Self {
        Self {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
            z: self.z - rhs.z,
        }
    }
}

impl SubAssign for Vector3D {
    #[inline]
    fn sub_assign(&mut self, rhs: Self) {
        self.x -= rhs.x;
        self.y -= rhs.y;
        self.z -= rhs.z;
    }
}

impl Mul for Vector3D {
    type Output = Self;
    #[inline]
    fn mul(self, rhs: Self) -> Self {
        Self {
            x: self.x * rhs.x,
            y: self.y * rhs.y,
            z: self.z * rhs.z,
        }
    }
}

impl MulAssign for Vector3D {
    #[inline]
    fn mul_assign(&mut self, rhs: Self) {
        self.x *= rhs.x;
        self.y *= rhs.y;
        self.z *= rhs.z;
    }
}

impl Div for Vector3D {
    type Output = Self;
    #[inline]
    fn div(self, rhs: Self) -> Self {
        Self {
            x: self.x / rhs.x,
            y: self.y / rhs.y,
            z: self.z / rhs.z,
        }
    }
}

impl DivAssign for Vector3D {
    #[inline]
    fn div_assign(&mut self, rhs: Self) {
        self.x /= rhs.x;
        self.y /= rhs.y;
        self.z /= rhs.z;
    }
}


impl Index<usize> for Vector3D {
    type Output = f64;
    fn index(&self, index: usize) -> &f64 {
        match index {
            0 => &self.x,
            1 => &self.y,
            2 => &self.z,
            _ => panic!("Index out of bounds"),
        }
    }
}

impl fmt::Display for Vector3D {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f,"{}i + {}j + {}k", self.x, self.y, self.z)
    }
}

impl fmt::Binary for Vector3D {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let magnitude = self.magnitude();
        let decimals = f.precision().unwrap_or(4);
        let string = format!("{magnitude:.decimals$}");
        f.pad_integral(true, "", &string)
    }
}

src/lib.rs
//! # i_mth
//!
//! This is a crate designed for use in easy to complex statics and dynamics problems.
//!
//! The crate is made as lean as possible and method names are directly correlated to what they
//! actually do.
//!
//! All types currently in the crate and features that will be implemented later will all use the
//! `f64` data type because well I am a freak of nature and I have accuracy issues.
//!
//!## Usage
//!
//!Add this to `Cargo.toml`
//!
//! ```toml
//! [dependencies]
//! i_mth = "0.1.2"
//! ```
//!
//! ## Current Crate available features
//!
//! 1. Support for both 2D and 3D vectors.
//! 1. Common Constants.
//! 1. A utils module for calculating escape velocity and acceleration of gravity of other celestial bodies. (other functions will be added in the future).
//!
//! ## Currently in the workings
//!
//! 1. Support For matrices, operations on matrices, rotations of vectors and vector transformations
//! 1. Numerical Calculus
//!
//! ### Examples
//!
//! 1. Creating 2D and 3D vectors with i_mth
//!
//! ```rust
//!    use i_mth::vector3d::Vector3D;
//!
//!    fn main() {
//!       // there are several available methods to create vectors
//!        let v3d_0 = Vector3D::new(1.0, 1.0, 1.0);
//!        let v3d_1 = Vector3D::set(1.0);
//!
//!        assert_eq!(v3d_0, v3d_1);
//!    }
//! ```
//!
//! 2. Moment of a force About a **O**: The moment of a force **F** about  **O** can be defined as the vector product (cross product) of **r** and **F**.
//! Where **r** is the position vector between the point of application of the force to the fixed reference point **O**.
//! ie **Moment** = **r x F**
//!
//! ```rust
//!    use i_mth::vector3d::Vector3D;
//!
//!    fn main() {
//!        let f = Vector3D::new(400.0, 693.0, 0.0);
//!        let r = Vector3D::new(-0.2, 0.16, 0.0);
//!
//!        let moment = r.cross(f);
//!        assert_eq!(-202.6, moment.z);
//!    }
//! ```
//! 3. Using the utils to calculate [escape velocity](https://en.wikipedia.org/wiki/Escape_velocity#:~:text=More%20generally%2C%20escape%20velocity%20is,orbit%20(of%20any%20radius).): Escape velocity or escape speed is the minimum speed needed for a free, non-propelled object to escape from the gravitational influence of a primary body, thus reaching an infinite distance from it.
//!
//! ```rust
//!     use i_mth::utils::calc_escape_velocity;
//! 
//!     fn main() {
//!         let mass_of_moon = 7.342e22;
//!         let radius_of_moon = 1737.4e3;
//! 
//!         let escape_vel_of_moon = calc_escape_velocity(mass_of_moon, radius_of_moon);
//! 
//!         println!("{} km/s", escape_vel_of_moon / 1000.0);
//!     }
//! ```
//! 
//! ### Supports fmt
//! 
//! ```rust
//!     use i_mth::vector3d::Vector3D;
//! 
//!     fn main() {
//!         let unit_i = Vector3D::i();
//!         let unit_j = Vector3D::j();
//! 
//!         let unit_k = unit_i.cross(unit_j);
//! 
//!         println!("{}", unit_k);
//!     }
//! ```
//! 


pub mod utils;
pub mod constants;

mod vectors;
pub use vectors::vector2d;
pub use vectors::vector3d;